File: python/day08.py
from aoc import *
from math import prod


def viewing_distance(height, d):
    return next((i for i, h in enumerate(d, 1) if h >= height), inf)

def solve(file=8):
    horizontal = mapl(digits, read_input(file))
    vertical = transpose(horizontal)
    size = len(horizontal)
    visible_from_outside = 0
    highest_score = 0
    for y in range(size):
        for x in range(size):
            height = horizontal[y][x]
            row = horizontal[y]
            col = vertical[x]
            directions = (
                row[:x][::-1],
                row[x+1:],
                col[:y][::-1],
                col[y+1:]
            )
            viewing_distances = [viewing_distance(height, d) for d in directions]
            visible_from_outside += any(dist == inf for dist in viewing_distances)
            visible_trees = (min(len(d), v)
                             for v, d in zip(viewing_distances, directions))
            highest_score = max(highest_score, prod(visible_trees))
    return visible_from_outside, highest_score


print(solve())










def test():
    assert solve("08_test") == (21, 8)
    assert solve() == (1829, 291840)

File: clojure/day08.clj
(ns day08
  (:require aoc))


(defn visible-from-outside? [^long height directions]
  (reduce
   (fn [acc dir]
     (if (every? #(< ^long % height) dir)
       (reduced true)
       acc))
   false
   directions))

(defn viewing-distance [^long height direction]
  (reduce
   (fn [^long acc ^long h]
     (if (< h height)
       (inc acc)
       (reduced (min (count direction) (inc acc)))))
   0
   direction))

(defn go-through-forrest [height-map]
  (let [hor  height-map
        vert (aoc/transpose height-map)]
    (for [[^long y row] (map-indexed vector hor)
          [^long x col] (map-indexed vector vert)
          :let [dirs [(rseq (subvec row 0 x)) ; left
                      (subvec row (inc x))    ; right
                      (rseq (subvec col 0 y)) ; up
                      (subvec col (inc y))]   ; down
                height (row x)]]
      {:is-visible?  (visible-from-outside? height dirs)
       :scenic-score (transduce (map #(viewing-distance height %)) * dirs)})))

(defn solve
  ([] (solve 8))
  ([input]
   (let [height-map (->> (aoc/read-input input)
                         (mapv aoc/string->digits))
         results (go-through-forrest height-map)]
     [(->> results (aoc/count-if :is-visible?))
      (->> results (map :scenic-score) (reduce max))])))


(solve)
